airflow/jobs.py

```diff
@@ -98,25 +98,48 @@ class BaseJob(Base, LoggingMixin):
         Index('idx_job_state_heartbeat', state, latest_heartbeat),
     )
 
+    heartrate = conf.getfloat('scheduler', 'JOB_HEARTBEAT_SEC'),
+
     def __init__(
             self,
             executor=executors.get_default_executor(),
-            heartrate=conf.getfloat('scheduler', 'JOB_HEARTBEAT_SEC'),
+            heartrate=None,
             *args, **kwargs):
         self.hostname = get_hostname()
         self.executor = executor
         self.executor_class = executor.__class__.__name__
         self.start_date = timezone.utcnow()
         self.latest_heartbeat = timezone.utcnow()
-        self.heartrate = heartrate
+        if heartrate is not None:
+            self.heartrate = heartrate
         self.unixname = getpass.getuser()
         self.max_tis_per_query = conf.getint('scheduler', 'max_tis_per_query')
         super(BaseJob, self).__init__(*args, **kwargs)
 
-    def is_alive(self):
+    @classmethod
+    @provide_session
+    def most_recent_job(cls, session):  # type: (Session) -> Optional[BaseJob]
+        """
+        Return the most recent job of this type, if any, based on last
+        heartbeat received.
+
+        This method should be called on a subclass (i.e. on SchedulerJob) to
+        return jobs of that type.
+
+        :rtype: BaseJob or None
+        """
+        return session.query(cls).order_by(cls.latest_heartbeat.desc()).limit(1).first()
+
+    def is_alive(self, grace_multiplier=2.1):
+        """
+        Is this job currently alive.
+
+        We define alive as in a state of RUNNING, and within a multiple of the
+        heartrate (default of 2.1)
```

Comment:
Maybe add some documentation about the the `grace_multiplier` in a `:param` + `:type` and the `:rtype`.

Reply:
Also do you think that it is more important for the function to execute fast (because of the use of `.utcnow()`) or could we split this expression into multiple expressions to become more readable?

Reply:
+1 on an explanation on why `2.1` is chosen as default

Reply:
I didn't actually change this function, I just added a doc string, but I don't think anything in the code base used this function before.

Reply:
The closest "alternative" was the scheduler_health_check_threshold parameter for the previous implementation of the /health endpoint, which had a default value of 30s (for a heart rate of 5s. so `multiplier=6`.)

I could change this parameter to a fixed time period, though a multiplier works better for this being in the base class as each of the Job types might have a different heartrate.